language/move-stdlib/sources/type_name.move

```diff
@@ -0,0 +1,29 @@
+/// Functionality for converting Move types into values. Use with care!
+module std::type_name {
+    use std::ascii::String;
+
+    struct TypeName has copy, drop, store {
+        /// String representation of the type. All types are represented
+        /// using their source syntax:
+        /// "u8", "u64", "u128", "bool", "address", "vector", "signer" for ground types.
+        /// Struct types are represented as fully qualified, `0x`-prefixed short type names
+        /// `0x1::string::String` or
+        /// `0xabcd::module_name1::type_name1<0x1234::module_name2::type_name2<u64>>`
+        /// Note: addresses in short type names are *always* shortened by stripping leading 0's
+        /// e.g., 0xA rather than 0x00000000A,
+        name: String
+    }
+
+    /// Return a value representation of the type `T`.
+    public native fun get<T>(): TypeName;
+
+    /// Get the String representation of `self`
+    public fun borrow_string(self: &TypeName): &String {
+        &self.name
```

Comment:
I am worried about exposing the `String` here, but if y'all are fine with fine by me. 

I think the root of my trepidation around the strings is it allows for cyclic thinking/checks with types. For example consider the difference between `t == TypeName::get<0x42::m::R>` vs `borrow_string(&t) == &ascii::new(b"0x42::m::R")`. The second can be used anywhere, where the first requires being able to link against `0x42::m`. 

But perhaps this cyclic style is desirable?

Reply:
Is your concern about both `borrow_string` and `into_string`, or only the former (guessing both, but wanted to double-check)?

I share the concern, and I think that the safest/most idiomatic usages of `TypeName`'s (and the ones easiest to verify with the prover) will look like your option (1). But I think hiding the string is a bit too paternalistic--if we're going to expose this form of reflection at all, I think we don't buy ourselves much by trying to limit its expressivity in this way.

Reply:
I don't see a problem in exposing a string. We have a very beautiful property specific to Move: the same 'name' (string) is unique for this type (and its layout) forever (assuming only compatible upgrades). There is no reason to hide the string. C++/Rust do this because they cannot guarantee uniqueness over different processes. We can.